File: source/go/internal/jyotish/planets.go
package jyotish

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/core"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var GrahaAbbr = [types.GrahaCount]string{
	types.GrahaSun:     "Su",
	types.GrahaMoon:    "Mo",
	types.GrahaMars:    "Ma",
	types.GrahaMercury: "Me",
	types.GrahaJupiter: "Ju",
	types.GrahaVenus:   "Ve",
	types.GrahaSaturn:  "Sa",
	types.GrahaRahu:    "Ra",
	types.GrahaKetu:    "Ke",
}

func MeanObliquity(t float64) float64 {
	return 23.439291111 -
		float64(0.013004167*t) -
		float64(0.000000164*t*t) +
		float64(0.000000504*t*t*t)
}

func isRetrograde(ctx *astronomy.EphemerisCtx, body astronomy.PlanetBody, ms int64) bool {
	const dt = 3600_000
	lon0 := astronomy.GetTropicalPlanetLongitude(ctx, body, ms-dt)
	lon1 := astronomy.GetTropicalPlanetLongitude(ctx, body, ms+dt)
	delta := lon1 - lon0
	if delta > 180 {
		delta -= 360
	}
	if delta < -180 {
		delta += 360
	}
	return delta < 0
}

func retrogradeDelta(ctx *astronomy.EphemerisCtx, body astronomy.PlanetBody, ms int64) float64 {
	const dt = 3600_000
	lon0 := astronomy.GetTropicalPlanetLongitude(ctx, body, ms-dt)
	lon1 := astronomy.GetTropicalPlanetLongitude(ctx, body, ms+dt)
	delta := lon1 - lon0
	if delta > 180 {
		delta -= 360
	}
	if delta < -180 {
		delta += 360
	}
	return delta
}

func getMeanRahuLongitudeTropical(ms int64) float64 {
	t := (astronomy.DateToJulianDay(ms) - 2451545.0) / 36525.0
	return utils.Normalize360(
		125.0445479 -
			float64(1934.1362891*t) +
			float64(0.0020754*t*t) +
			(t*t*t)/467441 -
			(t*t*t*t)/60616000,
	)
}

func getTrueRahuLongitudeTropical(ms int64) float64 {
	t := (astronomy.DateToJulianDay(ms) - 2451545.0) / 36525.0
	meanOmega := 125.0445479 -
		float64(1934.1362891*t) +
		float64(0.0020754*t*t) +
		(t*t*t)/467441 -
		(t*t*t*t)/60616000

	d := 297.8501921 +
		float64(445267.1114034*t) -
		float64(0.0018819*t*t) +
		(t*t*t)/545868 -
		(t*t*t*t)/113065000
	f := 93.2720950 +
		float64(483202.0175233*t) -
		float64(0.0036539*t*t) -
		(t*t*t)/3526000 +
		(t*t*t*t)/863310000

	argRad := ((float64(2*d) - float64(2*f)) * jsnum.PI) / 180
	correction := -1.4979 * math.Sin(argRad) // math.Sin, not the Cody-Waite astronomy.Sin
	return utils.Normalize360(meanOmega + correction)
}

func buildGrahaPosition(
	planet types.Graha,
	siderealLon float64,
	isRetro bool,
	nakshatraNameFn func(idx int) string,
	rashiNameFn func(idx int) string,
) types.GrahaPosition {
	rashiIndex := int(math.Floor(siderealLon / 30))
	degreeInRashi := siderealLon - float64(float64(rashiIndex)*30) // anti-FMA barrier
	nakIdx := utils.NakshatraOf(siderealLon)
	return types.GrahaPosition{
		Planet:            planet,
		SiderealLongitude: siderealLon,
		Rashi:             types.RashiInfo{Index: rashiIndex, Name: rashiNameFn(rashiIndex)},
		DegreeInRashi:     degreeInRashi,
		Nakshatra:         core.ComputeNakshatraFromLongitude(siderealLon, nakshatraNameFn(nakIdx)),
		IsRetrograde:      isRetro,
	}
}

func identityName(idx int) string { return jsnum.FormatInt(int64(idx)) }

var AllNodeTypes = []NodeType{NodeMean, NodeTrue}

func resolveNodeType(n NodeType) NodeType {
	if n == "" {
		return NodeMean
	}
	return n
}

func ComputePlanetaryPositions(
	ctx *astronomy.EphemerisCtx,
	ms int64,
	ayanamsaType types.AyanamsaType,
	nakshatraName func(idx int) string,
	rashiName func(idx int) string,
	nodeType NodeType,
) (types.PlanetaryPositions, error) {
	if nakshatraName == nil {
		nakshatraName = identityName
	}
	if rashiName == nil {
		rashiName = identityName
	}
	lon, err := siderealGrahaLongitudes(ctx, ms, ayanamsaType, nodeType)
	if err != nil {
		return types.PlanetaryPositions{}, err
	}

	marsRetro := isRetrograde(ctx, astronomy.PlanetMars, ms)
	mercRetro := isRetrograde(ctx, astronomy.PlanetMercury, ms)
	jupRetro := isRetrograde(ctx, astronomy.PlanetJupiter, ms)
	venRetro := isRetrograde(ctx, astronomy.PlanetVenus, ms)
	satRetro := isRetrograde(ctx, astronomy.PlanetSaturn, ms)

	g := func(planet types.Graha, retro bool) types.GrahaPosition {
		return buildGrahaPosition(planet, lon[planet], retro, nakshatraName, rashiName)
	}

	return types.PlanetaryPositions{
		Sun:     g(types.GrahaSun, false),
		Moon:    g(types.GrahaMoon, false),
		Mars:    g(types.GrahaMars, marsRetro),
		Mercury: g(types.GrahaMercury, mercRetro),
		Jupiter: g(types.GrahaJupiter, jupRetro),
		Venus:   g(types.GrahaVenus, venRetro),
		Saturn:  g(types.GrahaSaturn, satRetro),
		Rahu:    g(types.GrahaRahu, true), // always retrograde
		Ketu:    g(types.GrahaKetu, true),
	}, nil
}

// siderealGrahaLongitudes is the nine sidereal longitudes ComputePlanetaryPositions
// publishes, indexed by Graha, without its retrograde probes.
func siderealGrahaLongitudes(
	ctx *astronomy.EphemerisCtx,
	ms int64,
	ayanamsaType types.AyanamsaType,
	nodeType NodeType,
) ([types.GrahaCount]float64, error) {
	var lon [types.GrahaCount]float64
	nodeType = resolveNodeType(nodeType)

	ayanamsa, err := astronomy.ComputeAyanamsa(ms, ayanamsaType)
	if err != nil {
		return lon, err
	}

	toSidereal := func(tropical float64) float64 {
		return utils.Normalize360(tropical - ayanamsa)
	}

	sunSid, err := astronomy.GetSiderealSunLongitude(ctx, ms, ayanamsaType)
	if err != nil {
		return lon, err
	}

	moonSid, err := astronomy.GetSiderealMoonLongitude(ctx, ms, ayanamsaType)
	if err != nil {
		return lon, err
	}

	marsTrop := astronomy.GetTropicalPlanetLongitude(ctx, astronomy.PlanetMars, ms)
	mercTrop := astronomy.GetTropicalPlanetLongitude(ctx, astronomy.PlanetMercury, ms)
	jupTrop := astronomy.GetTropicalPlanetLongitude(ctx, astronomy.PlanetJupiter, ms)
	venTrop := astronomy.GetTropicalPlanetLongitude(ctx, astronomy.PlanetVenus, ms)
	satTrop := astronomy.GetTropicalPlanetLongitude(ctx, astronomy.PlanetSaturn, ms)

	var rahuTrop float64
	if nodeType == NodeTrue {
		rahuTrop = getTrueRahuLongitudeTropical(ms)
	} else {
		rahuTrop = getMeanRahuLongitudeTropical(ms)
	}
	ketuTrop := utils.Normalize360(rahuTrop + 180)

	lon[types.GrahaSun] = sunSid
	lon[types.GrahaMoon] = moonSid
	lon[types.GrahaMars] = toSidereal(marsTrop)
	lon[types.GrahaMercury] = toSidereal(mercTrop)
	lon[types.GrahaJupiter] = toSidereal(jupTrop)
	lon[types.GrahaVenus] = toSidereal(venTrop)
	lon[types.GrahaSaturn] = toSidereal(satTrop)
	lon[types.GrahaRahu] = toSidereal(rahuTrop)
	lon[types.GrahaKetu] = toSidereal(ketuTrop)
	return lon, nil
}

File: source/ts/src/jyotish/planets.ts
import { getTropicalPlanetLongitude, type PlanetBody } from '../astronomy/planet';
import { computeAyanamsa, dateToJulianDay } from '../astronomy/ayanamsa';
import { getSiderealSunLongitude } from '../astronomy/sun';
import { getSiderealMoonLongitude } from '../astronomy/moon';
import { computeNakshatraFromLongitude } from '../core/nakshatra';
import { normalize360 } from '../utils/angle';
import { nakshatraOf } from '../utils/constants';
import type { AyanamsaType } from '../types/options';
import type { GrahaPosition, GrahaName, PlanetaryPositions } from '../types/jyotish';

/** Two-letter abbreviations for the 9 grahas. */
export const GRAHA_ABBR: Record<GrahaName, string> = {
  Sun: 'Su', Moon: 'Mo', Mars: 'Ma', Mercury: 'Me',
  Jupiter: 'Ju', Venus: 'Ve', Saturn: 'Sa', Rahu: 'Ra', Ketu: 'Ke',
};

/** T in Julian centuries from J2000. */
export function meanObliquity(T: number): number {
  return 23.439291111 - 0.013004167 * T - 0.000000164 * T * T + 0.000000504 * T * T * T;
}

export { getTropicalPlanetLongitude };

function isRetrograde(body: PlanetBody, date: Date): boolean {
  const dt = 3600_000;
  const lon0 = getTropicalPlanetLongitude(body, new Date(date.getTime() - dt));
  const lon1 = getTropicalPlanetLongitude(body, new Date(date.getTime() + dt));
  let delta = lon1 - lon0;
  if (delta > 180) delta -= 360;
  if (delta < -180) delta += 360;
  return delta < 0;
}

/** Meeus *Astronomical Algorithms* 2nd ed. eq. 47.7; the mean node is classical Vedic practice for Rahu/Ketu. */
function getMeanRahuLongitudeTropical(date: Date): number {
  const T = (dateToJulianDay(date) - 2451545.0) / 36525.0;
  return normalize360(
    125.0445479
    - 1934.1362891 * T
    + 0.0020754 * T * T
    + (T * T * T) / 467441
    - (T * T * T * T) / 60616000,
  );
}

/** Only the dominant Meeus Ch. 47 term; the next (−0.1500°·sin M) and smaller are neglected. */
function getTrueRahuLongitudeTropical(date: Date): number {
  const T = (dateToJulianDay(date) - 2451545.0) / 36525.0;
  const meanΩ =
    125.0445479
    - 1934.1362891 * T
    + 0.0020754 * T * T
    + (T * T * T) / 467441
    - (T * T * T * T) / 60616000;

  const D =
    297.8501921
    + 445267.1114034 * T
    - 0.0018819 * T * T
    + (T * T * T) / 545868
    - (T * T * T * T) / 113065000;
  const F =
    93.2720950
    + 483202.0175233 * T
    - 0.0036539 * T * T
    - (T * T * T) / 3526000
    + (T * T * T * T) / 863310000;

  const argRad = ((2 * D - 2 * F) * Math.PI) / 180;
  const correction = -1.4979 * Math.sin(argRad);
  return normalize360(meanΩ + correction);
}

function buildGrahaPosition(
  planet: GrahaName,
  siderealLon: number,
  isRetro: boolean,
  nakshatraNameFn: (idx: number) => string,
  rashiNameFn: (idx: number) => string,
): GrahaPosition {
  const rashiIndex = Math.floor(siderealLon / 30);
  const degreeInRashi = siderealLon - rashiIndex * 30;
  const nakIdx = nakshatraOf(siderealLon);
  return {
    planet,
    siderealLongitude: siderealLon,
    rashi: { index: rashiIndex, name: rashiNameFn(rashiIndex) },
    degreeInRashi,
    nakshatra: computeNakshatraFromLongitude(siderealLon, nakshatraNameFn(nakIdx)),
    isRetrograde: isRetro,
  };
}

const identity = (idx: number) => String(idx);

/** Geocentric sidereal positions for all 9 grahas at `date`. */
export function computePlanetaryPositions(
  date: Date,
  ayanamsaType: AyanamsaType,
  nakshatraName: (idx: number) => string = identity,
  rashiName: (idx: number) => string = identity,
  nodeType: 'mean' | 'true' = 'mean',
): PlanetaryPositions {
  const ayanamsa = computeAyanamsa(date, ayanamsaType);

  const toSidereal = (tropical: number) => normalize360(tropical - ayanamsa);

  const sunSid = getSiderealSunLongitude(date, ayanamsaType);
  const moonSid = getSiderealMoonLongitude(date, ayanamsaType);

  const marsTrop = getTropicalPlanetLongitude('mars', date);
  const mercTrop = getTropicalPlanetLongitude('mercury', date);
  const jupTrop = getTropicalPlanetLongitude('jupiter', date);
  const venTrop = getTropicalPlanetLongitude('venus', date);
  const satTrop = getTropicalPlanetLongitude('saturn', date);

  const rahuTrop =
    nodeType === 'true' ? getTrueRahuLongitudeTropical(date) : getMeanRahuLongitudeTropical(date);
  const ketuTrop = normalize360(rahuTrop + 180);

  const marsRetro = isRetrograde('mars', date);
  const mercRetro = isRetrograde('mercury', date);
  const jupRetro = isRetrograde('jupiter', date);
  const venRetro = isRetrograde('venus', date);
  const satRetro = isRetrograde('saturn', date);

  const g = (planet: GrahaName, sid: number, retro: boolean) =>
    buildGrahaPosition(planet, sid, retro, nakshatraName, rashiName);

  return {
    sun:     g('Sun',     sunSid,                false),
    moon:    g('Moon',    moonSid,               false),
    mars:    g('Mars',    toSidereal(marsTrop),   marsRetro),
    mercury: g('Mercury', toSidereal(mercTrop),   mercRetro),
    jupiter: g('Jupiter', toSidereal(jupTrop),    jupRetro),
    venus:   g('Venus',   toSidereal(venTrop),    venRetro),
    saturn:  g('Saturn',  toSidereal(satTrop),    satRetro),
    rahu:    g('Rahu',    toSidereal(rahuTrop),   true),
    ketu:    g('Ketu',    toSidereal(ketuTrop),   true),
  };
}

